library/Fission/App/Creator/Class.hs

```diff
@@ -0,0 +1,46 @@
+module Fission.App.Creator.Class
+  ( Creator (..)
+  , Errors
+  ) where
+
+import           Database.Esqueleto (insert)
+import           Network.IPFS.CID.Types
+import           Servant
+
+import           Fission.Prelude
+import           Fission.Error as Error
+import           Fission.Models
+import           Fission.URL
+
+import qualified Fission.App.Domain   as AppDomain
+import           Fission.IPFS.DNSLink as DNSLink
+
+type Errors = OpenUnion
+  '[ ServerError
+   , AppDomain.AlreadyAssociated
+   , ActionNotAuthorized App
+   , NotFound            App
+   ]
+
+class Monad m => Creator m where
+  create :: UserId -> CID -> UTCTime -> m (Either Errors (AppId, Subdomain))
+
+instance (AppDomain.Initializer m, MonadDNSLink m) => Creator (Transaction m) where
+  create ownerId cid now = do
+    appId <- insert App
+      { appOwnerId    = ownerId
+      , appCid        = cid
+      , appInsertedAt = now
+      , appModifiedAt = now
+      }
+
+    AppDomain.associateDefault ownerId appId now >>= \case
```

Comment:
why don't you need an `Associator` monad as a constraint on this instance? It looks like it's a constraint to call `associateDefault`

Reply:
Actually I see this is the same thing for others as well (such as `AppDomain.allSiblingsByDomain` used in the `App.Destroyer` instance looking like it should have a `Retriever` constraint)

Reply:
> why don't you need an Associator monad as a constraint on this instance? It looks like it's a constraint to call associateDefault

Yeah, you'd think, right? The compiler tells you that it's not required here. Because they both have `MonadIO m =>` constraints, and it can resolve that they both work on _any_ `MonadIO` (think of it now as a property of MonadIO).

A simpler example is the concrete `Fission` type. It has instances for all of these classes, but they don't need special constraints, because Fission _is a member of_ each typeclass (it is an Associator, a Modifier, &c). The specific `Associator` constraint is thus redundant.

Typeclasses work the opposite direction from inheritance, more like a mixin —- it's superclassing rather than subclassing.

Reply:
ah got it. ok very cool :+1: